### PR TITLE
MOD-5739: Configurable minimum length of word to stem

### PR DESCRIPTION
**Pull request by nafraf**
**Description**

Expose `MINSTEMLEN` as configuration parameters

**Which issues this PR fixes**
1. MOD-5739

**Main objects this PR modified**
1. ...
2. ...

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

**Review by raz-mon — COMMENTED** [review]
Nice work! 💪 
See some small comments.

One concern I have here, is that if we allow modifying the minimal stemming length in runtime, we may open a door for some ambiguities for scenarios such as:
* User creates an index with the default minimal stemming length (4).
* User populates the database.
* User changes minimal stemming length to 3.
* User continues to populate the index.

Now there may be a confusing situation in which the user isn't sure what was indexed with which minimal stemming length, and therefore what to expect for his queries. Also, the `INFO` data corresponding to this value will not reflect its value throughout the whole index lifetime, which may also cause confusions.

My intuition when going over the PR was to allow the user to set the minimal stemming length at index creation time - only, such that there are no such ambiguities, and he can set it to different values per index which is nicer flexibility that what we have today.

We can also just say that it's the user's responsibility to track this. This seems a bit less user-friendly to me, even though it allows maximum flexibility, which is nice in a way.

File: src/config.c [hunk 1]
```diff
@@ -61,6 +61,17 @@ CONFIG_GETTER(getMinPrefix) {
   return sdscatprintf(ss, "%lld", config->iteratorsConfigParams.minTermPrefix);
 }
 
+// MINSTEMLEN
+CONFIG_SETTER(setMinStemLen) {
+  int acrc = AC_GetLongLong(ac, &config->iteratorsConfigParams.minStemLength, AC_F_GE1);
```
[the remark this answers is not in the document]
Comment (nafraf) [reply src/config.c]: OK. The validation was added.

File: tests/pytests/test_config.py [hunk 2]
```diff
@@ -164,6 +167,7 @@ def test_arg_num(arg_name, arg_value):
     test_arg_num('UNION_ITERATOR_HEAP', 20)
     test_arg_num('_NUMERIC_RANGES_PARENTS', 1)
     test_arg_num('BG_INDEX_SLEEP_GAP', 15)
+    test_arg_num('MINSTEMLEN', 3)
```
Comment (raz-mon) [inline tests/pytests/test_config.py:170]: Didn't you set its default to be `4` 🤔 ?

File: src/config.c [hunk 3]
```diff
@@ -61,6 +61,23 @@ CONFIG_GETTER(getMinPrefix) {
   return sdscatprintf(ss, "%lld", config->iteratorsConfigParams.minTermPrefix);
 }
 
+// MINSTEMLEN
+CONFIG_SETTER(setMinStemLen) {
+  long long minStemLen;
+  int acrc = AC_GetLongLong(ac, &minStemLen, AC_F_GE1);
+  if (minStemLen < MIN_MIN_STEM_LENGHT) {
+    QueryError_SetErrorFmt(status, MIN_MIN_STEM_LENGHT, "Minimum stem length cannot be lower than %lld", MIN_MIN_STEM_LENGHT);
```
Comment (raz-mon) [inline src/config.c:69]: Consider using `AC_ERR_ELIMIT`, which is used for most config-param bounds errors.
This is more explicit which is nice too, but less coherent with the existing similar errors.
Your call.

**Review by raz-mon — COMMENTED** [review]
Nice.
See one comment

File: tests/pytests/test_config.py (hunk 2 above)
Comment (nafraf) [reply tests/pytests/test_config.py:170]: Yes. But `test_arg_num` is not using the default value, it is passing  `MINSTEMLEN 3` as module argument

File: src/config.c (hunk 3 above)
Comment (nafraf) [reply src/config.c]: OK. Change done. Thanks.

**Review by raz-mon — APPROVED** [review]
LGTM 👍

File: src/config.c [hunk 4]
```diff
@@ -61,6 +61,22 @@ CONFIG_GETTER(getMinPrefix) {
   return sdscatprintf(ss, "%lld", config->iteratorsConfigParams.minTermPrefix);
 }
 
+// MINSTEMLEN
+CONFIG_SETTER(setMinStemLen) {
+  long long minStemLen;
+  int acrc = AC_GetLongLong(ac, &minStemLen, AC_F_GE1);
+  if (minStemLen < MIN_MIN_STEM_LENGHT) {
+    RETURN_STATUS(AC_ERR_ELIMIT);
```
Comment (oshadmi) [inline src/config.c:69]: Should notify the user about the limit

File: tests/pytests/test_config.py [hunk 5]
```diff
@@ -17,6 +17,8 @@ def testConfigErrors(env):
     env.expect('ft.config', 'idx').error().contains('wrong number of arguments')
     env.expect('ft.config', 'set', '_NUMERIC_RANGES_PARENTS', 3) \
         .equal('Max depth for range cannot be higher than max depth for balance')
+    env.expect('ft.config', 'set', 'MINSTEMLEN', 1).error()\
+        .contains('Value is outside acceptable bounds')
```
Comment (oshadmi) [inline tests/pytests/test_config.py:21]: Should reveal the value `MIN_MIN_STEM_LENGHT`

File: tests/pytests/test_stemmer.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+    env.cmd('HSET', '{doc}:3', 't', 'frying')
+    res = env.cmd('FT.SEARCH', 'idx_min3', 'fried', 'SORTBY', 't', 'ASC')
+    env.assertEqual(res, [3, '{doc}:2', ['t', 'fries'], '{doc}:1', ['t', 'fry'],
+                          '{doc}:3', ['t', 'frying']])
+    
+    res = env.cmd(config_cmd(), 'GET', 'MINSTEMLEN')
+    env.assertEqual(res[0], ['MINSTEMLEN', '3'])
+
+    ########################################################
+    # Test the default MIN_STEMMING_LEN = 5
```
[the remark this answers is not in the document]
Comment (nafraf) [reply tests/pytests/test_stemmer.py]: OK. The test with `MIN_STEMMING_LEN = 5 was removed`

File: tests/pytests/test_stemmer.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    env.cmd('JSON.SET', '{doc}:3', '$', r'{"t":"frying"}')
+    res = env.cmd('FT.SEARCH', 'idx_min3', 'fried', 'SORTBY', 't', 'ASC',
+                  'NOCONTENT')
+    env.assertEqual(res, [3, '{doc}:2', '{doc}:1', '{doc}:3'])
+    
+    res = env.cmd(config_cmd(), 'GET', 'MINSTEMLEN')
+    env.assertEqual(res[0], ['MINSTEMLEN', '3'])
+
+    ########################################################
+    # Test the default MIN_STEMMING_LEN = 5
```
[the remark this answers is not in the document]
Comment (nafraf) [reply tests/pytests/test_stemmer.py]: OK. The test with `MIN_STEMMING_LEN = 5 was removed`

File: src/config.h [hunk 8]
```diff
@@ -243,6 +247,7 @@ void DialectsGlobalStats_AddToInfo(RedisModuleInfoCtx *ctx);
     .extLoad = NULL,                                                                                                  \
     .gcConfigParams.enableGC = 1,                                                                                     \
     .iteratorsConfigParams.minTermPrefix = 2,                                                                         \
+    .iteratorsConfigParams.minStemLength = 4,                                                                         \
```
[the remark this answers is not in the document]
Comment (nafraf) [reply src/config.h]: Done. [7a0082b](https://github.com/RediSearch/RediSearch/pull/4587/commits/7a0082b86b52fc773d07832a7c844f960bce896e)

File: tests/pytests/test_config.py (hunk 5 above)
Comment (nafraf) [reply tests/pytests/test_config.py]: Fixed.

**Review by oshadmi — APPROVED** [review]

**Pull request merged**